File: packages/elixir/examples/visitor_basic.exs
#!/usr/bin/env elixir
# Basic Visitor Pattern Example
# Demonstrates core visitor functionality

defmodule BasicVisitor do
  @moduledoc """
  A simple visitor that tracks which callbacks are invoked.
  """

  use HtmlToMarkdown.Visitor

  @impl true
  def handle_heading(_context, level, text, _id) do
    IO.puts("  Found heading level #{level}: #{text}")
    :continue
  end

  @impl true
  def handle_link(_context, href, text, _title) do
    IO.puts("  Found link: #{text} -> #{href}")
    :continue
  end

  @impl true
  def handle_image(_context, src, alt, _title) do
    IO.puts("  Found image: #{alt} (#{src})")
    :continue
  end
end

# Example HTML
html = """
<html>
<head><title>Example</title></head>
<body>
  <h1>Welcome</h1>
  <p>This is a <a href="https://example.com">link</a> and an <img src="test.jpg" alt="Test Image"></p>
  <h2>Section</h2>
  <p>More content with <a href="https://test.com">another link</a>.</p>
</body>
</html>
"""

IO.puts("Converting HTML with BasicVisitor:")
IO.puts("")

{:ok, markdown} = HtmlToMarkdown.Visitor.convert_with_visitor(html, BasicVisitor, nil)

IO.puts("")
IO.puts("Converted Markdown:")
IO.puts("---")
IO.puts(markdown)
IO.puts("---")

File: packages/php/examples/visitor_basic.php
<?php

declare(strict_types=1);

/**
 * Basic Visitor Example
 *
 * This example demonstrates how to use the visitor pattern to customize
 * HTML-to-Markdown conversion by intercepting and modifying how certain
 * elements are processed.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use HtmlToMarkdown\Visitor\AbstractVisitor;
use HtmlToMarkdown\Visitor\NodeContext;
use HtmlToMarkdown\Visitor\VisitResult;
use HtmlToMarkdown\HtmlToMarkdown;

/**
 * Custom visitor that wraps all links in extra brackets for emphasis
 */
class LinkHighlighter extends AbstractVisitor
{
    public function visitLink(NodeContext $context, string $href, string $text, ?string $title): array
    {
        return VisitResult::custom(">> [{$text}]({$href}) <<");
    }
}

$html = <<<'HTML'
<article>
    <h1>Welcome to My Blog</h1>
    <p>Check out <a href="https://example.com">this example</a> and <a href="https://test.com">this one too</a>.</p>
</article>
HTML;

echo "Default conversion:\n";
echo HtmlToMarkdown::convert($html);
echo "\n---\n\n";

echo "With custom visitor (if supported):\n";
try {
    $visitor = new LinkHighlighter();
    $markdown = HtmlToMarkdown::convertWithVisitor($html, null, $visitor);
    echo $markdown;
} catch (Throwable $e) {
    echo "Note: Visitor example requires the PHP extension with visitor support\n";
    echo "Error: " . $e->getMessage() . "\n";
}
